### DISCUSSION TITLE
Need help with jdtls configuration with lombok

### DISCUSSION THREAD

**Discussion by titoOdUA** [LazyVim/LazyVim#275] [Q&A] [answered]
Hello,
Can someone please help me with jdtls config?
Is there an example of how to configure java+lombok lsp with LazyVim?

Basically, I force it to work with the help of jdtls-launcher.
But for some reason, lombok is not working.

I have tried to download lombok.jar manually and specify the absolute path in the launcher script but it didn't help.

I am aware that the question is hardly about LazyVim, but it is the last thing that prevents the switch for me (if it is against the rules and must be deleted it is ok, sorry for the trouble).

I am relatively new to nvim in general.
Currently switched to lazyVim for Rust development (works like a charm), but can't wrap my head around the java+lombok configuration.
Would be grateful for some kind of help or direction.

**Comment by titoOdUA** [comment, marked as answer]
Well, finally made it work, but with another nvim config repo. Will try to replicate the behavior in LazyVim (so far no luck).
Probably missing something.
Do not need help anymore =)

**Reply by Kim-Robin** [reply]
Hi @titoOdUA. I am having trouble setting up java with LazyVim.(I am new to lua and neovim). If possible can you share your configuration for java and lombok please?

**Reply by guillaumeLamanda** [reply]
Here a working config with lombok. It still misses the debugging configuration, I struggle to configure it 

```lua
return {
  "neovim/nvim-lspconfig",
  dependencies = {
    "mfussenegger/nvim-jdtls",
  },
  init = function()
    require("lazyvim.util").on_attach(function(client)
      -- if client.name == "jdtls" then
      --   local jdtls = require("jdtls")
      --   jdtls.setup_dap({ hotcodereplace = "auto" })
      --   jdtls.setup.add_commands()
      --   -- Auto-detect main and setup dap config
      --   require("jdtls.dap").setup_dap_main_class_configs({
      --     config_overrides = {
      --       -- vmArgs = "-Dspring.profiles.active=local",
      --     },
      --   })
      -- end
    end)
  end,
  opts = {
    -- LSP Server Settings
    --- @type lspconfig.options
    setup = {
      jdtls = function(_, opts)
        -- calculate workspace dir
        local project_name = vim.fn.fnamemodify(vim.fn.getcwd(), ":p:h:t")
        local workspace_dir = vim.fn.stdpath("data") .. "/site/java/workspace-root/" .. project_name
        os.execute("mkdir " .. workspace_dir)
        local jdtls = require("jdtls")

        -- get the mason install path
        local mason_registry = require("mason-registry")
        local install_path = require("mason-registry").get_package("jdtls"):get_install_path()
        local java_test_install_path = require("mason-registry").get_package("java-test"):get_install_path()

        -- get the current OS
        local os
        if vim.fn.has("mac") == 1 then
          os = "mac"
        elseif vim.fn.has("unix") == 1 then
          os = "linux"
        elseif vim.fn.has("win32") == 1 then
          os = "win"
        end

        local bundles = {
          mason_registry.get_package("java-debug-adapter"):get_install_path()
            .. "/extension/server/com.microsoft.java.debug.plugin-0.45.0.jar",
        }
        vim.list_extend(
          bundles,
          vim.split(
            vim.fn.glob(mason_registry.get_package("java-test"):get_install_path() .. "/extension/server/*.jar"),
            "\n"
          )
        )

        -- vim.list_extend(bundles, vim.split(vim.fn.glob("/home/jemag/dev/vscode-java-test/server/*.jar"), "\n"))
        -- ensure that OS is valid
        if not os or os == "" then
          vim.notify("jdtls: Could not detect valid OS", vim.log.levels.ERROR)
        end
        local jdtls_opts = vim.tbl_deep_extend("force", opts, {
          cmd = {
            "java",
            "-Declipse.application=org.eclipse.jdt.ls.core.id1",
            "-Dosgi.bundles.defaultStartLevel=4",
            "-Declipse.product=org.eclipse.jdt.ls.core.product",
            "-Dlog.protocol=true",
            "-Dlog.level=ALL",
            "-javaagent:" .. install_path .. "/lombok.jar",
            "-Xms1g",
            "--add-modules=ALL-SYSTEM",
            "--add-opens",
            "java.base/java.util=ALL-UNNAMED",
            "--add-opens",
            "java.base/java.lang=ALL-UNNAMED",
            "-jar",
            vim.fn.glob(install_path .. "/plugins/org.eclipse.equinox.launcher_*.jar"),
            "-configuration",
            install_path .. "/config_" .. os,
            "-data",
            workspace_dir,
          },
          init_options = {
            bundles = bundles,
          },
          filetypes = { "java" },
        })
        require("lazyvim.plugins.lsp")
        jdtls.start_or_attach(jdtls_opts)
        return true
      end,
    },
  },
}
```

**Reply by Kim-Robin** [reply]
@guillaumeLamanda do you have public repo for lazyvim that I can take a look?

**Reply by guillaumeLamanda** [reply]
Nope, but from [this](https://github.com/LazyVim/LazyVim/discussions/42#discussion-4746848) discussion I found [this](https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/extras/lang/java.lua) configuration that works really great

**Reply by Kim-Robin** [reply]
I have looked into that one before but I have some problem with "go to definition". It gives error saying 

` Error executing vim.schedule lua callback: ...9.1/nvim-linux64/share/nvim/runtime/lua/vim/lsp/util.lua:1146: BufReadCmd Autocommands for "jdt://*": Vim(lua):Error executing vim.schedule lua callback: /home/login/.local/share/nvim/lazy/nvim-jdtls/lua/jdtls.lua:1163: Vim:Error executing vim.schedule lua callback: ...e/nvim/lazy/lualine.nvim/lua/lualine/utils/nvim_opts.lua:77: E539: Illegal character </>
stack traceback:
	[C]: in function 'nvim_win_set_option'
	...e/nvim/lazy/lualine.nvim/lua/lualine/utils/nvim_opts.lua:77: in function 'setter_fn'
	...e/nvim/lazy/lualine.nvim/lua/lualine/utils/nvim_opts.lua:50: in function 'set_opt'
	...e/nvim/lazy/lualine.nvim/lua/lualine/utils/nvim_opts.lua:74: in function 'set'
	...ogin/.local/share/nvim/lazy/lualine.nvim/lua/lualine.lua:435: in function 'refresh'
	...ogin/.local/share/nvim/lazy/lualine.nvim/lua/lualine.lua:356: in function <...ogin/.local/share/nvim/lazy/lualine.nvim/lua/lualine.lua:355>
	[C]: in function 'wait'
	/home/login/.local/share/nvim/lazy/nvim-jdtls/lua/jdtls.lua:1181: in function 'open_classfile'
	[string ":lua"]:1: in main chunk
	[C]: in function 'nvim_win_set_buf'
	...9.1/nvim-linux64/share/nvim/runtime/lua/vim/lsp/util.lua:1146: in function 'jump_to_location'
	...nvim/lazy/telescope.nvim/lua/telescope/builtin/__lsp.lua:200: in function 'handler'
	...b/v0.9.1/nvim-linux64/share/nvim/runtime/lua/vim/lsp.lua:1394: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>
stack traceback:
	[C]: in function 'nvim_buf_set_lines'
	/home/login/.local/share/nvim/lazy/nvim-jdtls/lua/jdtls.lua:1163: in function 'handler'
	...b/v0.9.1/nvim-linux64/share/nvim/runtime/lua/vim/lsp.lua:1394: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>
`

**Reply by guillaumeLamanda** [reply]
@titoOdUA here is a working example using the LazyVim java extra : 

```lua
return {
  {
    "mfussenegger/nvim-jdtls",
    ---@type lspconfig.options.jdtls
    ---@diagnostic disable-next-line: missing-fields
    opts = {
      jdtls = function(opts)
        local install_path = require("mason-registry").get_package("jdtls"):get_install_path()
        local jvmArg = "-javaagent:" .. install_path .. "/lombok.jar"
        table.insert(opts.cmd, "--jvm-arg=" .. jvmArg)
        return opts
      end,
    },
  },
}

```

**Comment by Frederick888** [comment]
This was pure ignorance, but those were a few hours that I can never get back, so I'll drop a pointer here...

I was migrating from coc.nvim and I wanted to use my local copy of jdt.ls with some tweaks.

Note that if you build the `cmd` from scratch, you MUST put `-javaagent:` before `-jar` (so that it's picked up as a JVM arg)! For example,
```lua
local java_agents = vim.g.java_agents or {}
java_agents = vim.tbl_map(function(agent)
  return '-javaagent:' .. vim.fs.normalize(agent)
end, java_agents)

local cmd = {
  java_bin,
  '-Declipse.application=org.eclipse.jdt.ls.core.id1',
  '-Dosgi.bundles.defaultStartLevel=4',
  '-Declipse.product=org.eclipse.jdt.ls.core.product',
  -- '-Dlog.level=ALL',
  -- '-Dlog.protocol=true',
  '-Dsun.zip.disableMemoryMapping=true',
  '-Dfile.encoding=UTF-8',
  '-Djava.import.generatesMetadataFilesAtProjectRoot=false',
  '-Djdt.core.sharedIndexLocation=' .. shared_index,
  '-Xms' .. java_init_alloc_pool,
  '-Xmx' .. java_max_alloc_pool,
  '-XX:+HeapDumpOnOutOfMemoryError',
  '-XX:HeapDumpPath=' .. data_root .. '/_dumps',
  '-XX:+UseParallelGC',
  '-XX:GCTimeRatio=4',
  '-XX:AdaptiveSizePolicyWeight=90',
  '-Xlog:disable',
  '--add-modules=ALL-SYSTEM',
  '--add-opens', 'java.base/java.util=ALL-UNNAMED',
  '--add-opens', 'java.base/java.lang=ALL-UNNAMED',
  '--add-opens', 'java.base/sun.nio.fs=ALL-UNNAMED',
}
vim.list_extend(cmd, java_agents)
vim.list_extend(cmd, {
  '-jar', jdtls_launcher,
  '-configuration', jdtls_config,
  '-data', project_data,
})
return cmd
```

...instead of `vim.list_extend(cmd, java_agents)` at the end.

This is also why when using the `bin/jdtls[.py]` wrapper, `--jvm-arg=-javaagent:...` works but only `-javaagent:` doesn't:
```py
exec_args = ["-Declipse.application=org.eclipse.jdt.ls.core.id1",
                "-Dosgi.bundles.defaultStartLevel=4",
                "-Declipse.product=org.eclipse.jdt.ls.core.product",
                "-Dosgi.checkConfiguration=true",
                "-Dosgi.sharedConfiguration.area=" + str(shared_config_path),
                "-Dosgi.sharedConfiguration.area.readOnly=true",
                "-Dosgi.configuration.cascaded=true",
                "-Xms1G",
                "--add-modules=ALL-SYSTEM",
                "--add-opens", "java.base/java.util=ALL-UNNAMED",
                "--add-opens", "java.base/java.lang=ALL-UNNAMED"] \
                + known_args.jvm_arg \  # it works here
                + ["-jar", jar_path,
                "-data", known_args.data] \
                + args  # but not here
```

**Reply by Konfekt** [reply]
Thank you for the heads up.

What about setting the `$JDTLS_JVM_ARGS = '-javaagent:' .. $HOME .. '/lombok.jar'` ? Does Jtdls then still require  `--jvm-arg=$JDTLS_JVM_ARGS`?
